### PR TITLE
MLP implementation in PyTorch & suggests a new schema for PyTorch implementations w.r.t Issue #8699

### PR DESCRIPTION
**Pull request by RecreationalMath**
## Summary
Draft: WIP on Issue #8699 , sharing here to gather feedback. **Suggestions are welcome**

- This PR suggest an alternative and a modular way of setting up network construction for PyTorch models in sktime. 
- Implements new MLP torch model for classification.
- I have tried to adhere to the design and flow structure that is already in place for deep learning methods implemented in TensorFlow in sktime. 
- Other WIPs which I have referenced and tried to improve upon are #8766 ,  #8712  and #8710 

### 1. Base PyTorch network class implementation & design considerations

Presently, there is a `/sktime/classification/deep_learning/_pytorch.py` file which handles PyTorch network creation.
Issues & suggestions regarding this file:

1. This single file residing in `/sktime/classificaton/deep_learning/` folder has the common PyTorch classes and methods which will get used for both, classification and regression task but is defined within classification directory. Whereas all such TensorFlow network classes and method signatures are declared in `/sktime/networks/` folder. Following the same structure for PyTorch will make the codebase more manageable and modular.
2. For all TensorFlow models, they are abstracted as follows -> 

> 1. There is a Base Network class which is common for all different model architectures and gets inherited by them.
> 2. Then there is a common model architecture class for both classification & regression tasks and gets inherited by them.
> 3. Finally there is a separate class defining model architecture for each task, classification and regression.

`_pytorch.py` file combines abstraction level 1 and 2 in a single file/class. Following same structure as for TensorFlow model will make the PyTorch models more manageable and easily understandable for any developer working on both frameworks.

3. This file doesn't follow some of the standard pytorch practices, for example: every network class in PyTorch should inherit from torch.nn.Module. Following the standard pytorch practices will make the codebase more readable and easier for new sktime developers to contribute.

### 2. Checking for PyTorch dependencies

`_check_dl_dependencies` method defined in `/sktime/utils/dependencies/_dependencies.py` only checks for TensorFlow package. Instead of using the same method to check for torch package, I have defined a new `_check_torch_dependencies` method. 
**Reasoning**: The previous function often gets called without package/module name as an argument. Upon reading the above mentioned `_dependencies.py` file it seemed as a design consideration to do it this way. PyTorch currently gets checked as a soft dependency and any module/package wise dependency checking seemed to be done through their individual methods. Hence, the new method.

PS: I need help here as I am more familiar with requirements.txt way of doing things.

### 3. MLP model building for classification, in PyTorch.
Currently working on implementing the MLP model for classification using the newly created pytorch network classes and methods.

## Code Quality
- Passes all pre-commit hooks including ruff formatting and linting.
- Will check thoroughly to match sktime guidlines once the implementation is completed.

## Testing Infrastructure
To be done

## TensorFlow to PyTorch migration checks
To be done

Check for same defaults in both, tensorflow and torch models, w.r.t:
- [ ] loss functions
- [ ] activation functions
- [ ] dropouts
- [ ] hidden layer size and dimension of each layer
- [ ] optimizers

### CODE REVIEW

File: sktime/utils/dependencies/_dependencies.py [hunk 1]
```diff
@@ -315,6 +315,45 @@ def _check_dl_dependencies(msg=None, severity="error"):
         return False
 
 
+def _check_torch_dependencies(msg=None, severity="error"):
```
Comment (fkiraly) [inline sktime/utils/dependencies/_dependencies.py]: I think you do not need this, you can use `_check_soft_dependencies` directly? `_check_dl_dependencies` is just a relic from times before we had the general feature

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks for the ideas!

Some remarks:

* soft dependencies are not isolated, and the module `mlp_torch` will fail when `torch` is not present - see here: https://www.sktime.net/en/latest/developer_guide/dependencies.html#handling-soft-dependencies
* could you kindly coordinate with @jnrusson1 if you are working on the same thing? He has a more complete design here: https://github.com/sktime/sktime/pull/8766
    * there is more than enough work on the neural networks! So there is no need to work on exactly the same issue.
* how would this look like together with the classifier?

File: sktime/utils/dependencies/_dependencies.py (hunk 1 above)
Comment (RecreationalMath) [reply sktime/utils/dependencies/_dependencies.py]: Ok, got it. 
Will remove this in the next commit.

**Comment by RecreationalMath** [comment]
1. Thanks for pointing to [Handling the soft dependencies](https://www.sktime.net/en/latest/developer_guide/dependencies.html#handling-soft-dependencies), this helps.  I was looking for something like this in the docs.
2. Yeah, I was working on RNN initially, but then saw [#8766](https://github.com/sktime/sktime/pull/8766), and I wanted to showcase an alternate way of doing what he was trying to do. Instead of leaving a ton of comments and going back-n-forth, thought it might be easier for us to discuss if I just show the interfaces I had in mind on the most basic network structure and list my reasoning in one place, hence the Draft PR. Want to get the interfaces right, no need to merge this.
3. I'm working on implementing the classifier next, shall commit soon.

**Comment by RecreationalMath** [comment]
@fkiraly I was under the impression that we have to follow API design mentioned [here](https://github.com/sktime/pytorch-forecasting/issues/1736) for the PyTorch models. 

Following from the discussion in comments there, I was borrowing interface design from [MLP design in pytorch-forecasting](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/mlp/submodules.py), [Base class design from DSIPTS](https://github.com/DSIP-FBK/DSIPTS/blob/main/dsipts/src/dsipts/models/base.py) and [RNN model design from DSIPTS](https://github.com/DSIP-FBK/DSIPTS/blob/main/dsipts/src/dsipts/models/RNN.py)
Perhaps that's why my interface understanding was very different from #8766 and led me to start working on an alternate PR for MLP, as the latter strongly follows the already implemented TensorFlow interfaces. 

Could you please clarify which API design we have to follow?

Hi @jnrusson1 👋, 
Would be helpful if you could please comment on your approach. Thanks!

**Comment by agobbifbk** [comment]
> @fkiraly I was under the impression that we have to follow API design mentioned [here](https://github.com/sktime/pytorch-forecasting/issues/1736) for the PyTorch models.
> 
> Following from the discussion in comments there, I was borrowing interface design from [MLP design in pytorch-forecasting](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/mlp/submodules.py), [Base class design from DSIPTS](https://github.com/DSIP-FBK/DSIPTS/blob/main/dsipts/src/dsipts/models/base.py) and [RNN model design from DSIPTS](https://github.com/DSIP-FBK/DSIPTS/blob/main/dsipts/src/dsipts/models/RNN.py) Perhaps that's why my interface understanding was very different from #8766 and led me to start working on an alternate PR for MLP, as the latter strongly follows the already implemented TensorFlow interfaces.
> 
> Could you please clarify which API design we have to follow?
> 
> Hi @jnrusson1 👋, Would be helpful if you could please comment on your approach. Thanks!

Hello there my two cents:
1- Why there is a flatten layer at the beginning? What in the input shape?
2- Avoid using activation function at the end, unless there is a particular reason to do that
3- Is this a classification task or a regression task? What is the shape of the target variable? You may want to reshape also the output
4- Since all this lives inside `sktime` I suggest to follow the `sktime` API, not the DSIPTS one

If you have some technical questions PM me on discord! 
Have a nice day!

**Comment by RecreationalMath** [comment]
Thanks for the comments. Closing this. 
Took notes from this and created a new PR for the RNN model.
